File: apps/composites/assign-job/src/consumer.py
import json
import logging

import aio_pika
from townops_shared.utils.rabbitmq import RabbitMQClient

from .services import assign_contractor

logger = logging.getLogger(__name__)

client = RabbitMQClient()


async def process_case_opened_message(body: dict) -> None:
  try:
    await assign_contractor(
      case_id=body["case_id"],
      postal_code=body["postal_code"],
      category_code=body["category_code"],
    )
    logger.info("Assignment created for case %s", body["case_id"])
  except Exception:
    logger.exception("Failed to process Case_Opened")
    raise


async def _on_message(message: aio_pika.abc.AbstractIncomingMessage) -> None:
  body = json.loads(message.body)
  await process_case_opened_message(body)


async def start_consumer() -> None:
  await client.connect()
  await client.consume(queue_name="case.opened", callback=_on_message)

File: apps/composites/assign-job/src/consumer.ts
import { logger } from "@townops/shared-ts";

import { assignContractor } from "./services";

export interface CaseOpenedEvent {
  caseId: string;
  residentId: string;
  category: string;
  priority: string;
  postalCode?: string;
  addressDetails?: string;
  description?: string;
}

/**
 * Handler for case.opened events consumed from the assign-job-queue.
 * Orchestrates contractor selection and assignment creation.
 */
export async function handleCaseOpened(message: any): Promise<void> {
  let event: CaseOpenedEvent;

  try {
    const raw =
      typeof message === "string"
        ? message
        : (message.bodyString?.() ?? JSON.stringify(message));
    event = JSON.parse(raw) as CaseOpenedEvent;
  } catch {
    logger.error(
      { message },
      "assign-job: failed to parse case.opened message"
    );
    return;
  }

  if (!event.caseId) {
    logger.error(
      { event },
      "assign-job: invalid case.opened payload — missing caseId"
    );
    return;
  }

  const postalCode =
    event.postalCode ?? event.addressDetails?.slice(0, 6) ?? "000000";
  const categoryCode = event.category;

  logger.info(
    { caseId: event.caseId, postalCode, categoryCode },
    "assign-job: processing case.opened event"
  );

  await assignContractor(event.caseId, postalCode, categoryCode);
}
